accepted/00000-replace-salt-toaster.md

```diff
@@ -0,0 +1,192 @@
+
+## Overview
+
+The Salt upstream test suite is executed in virtual machines (VMs) that run the
+different operating systems we support. These VMs are set up with sumaform and
+salt-jenkins. Terracumber triggers terraform to create the VMs from within a
+Jenkins job and launches the test suite, connecting to each VM over SSH.
+
+## Provisioning Test VMs with Sumaform + Salt-Jenkins
+
+A new module "salt\_testenv" is added to sumaform. This module defines a new
+role, "salt\_testenv", and the default values needed to run all test suites, e.g.
+4GiB of RAM. Sumaform uses this role to apply the `golden-images-provision.sls`
+state file from the Salt-Jenkins repository.
+
+Up to now, all state files used by sumaform are maintained in the sumaform Git
+repository. For provisioning the salt\_testenv VMs, we don't want to maintain
+our own states that define which dependencies need to be installed on what
+operating system. Salt upstream already maintains such states for their own test
+infrastructure. We re-use their states by adding `salt-jenkins` as a Git
+submodule to sumaform and merge the salt state trees.
+
+In order to avoid future state id clashes, all existing state ids will get a
+"sumaform-" prefix.
+
+### Merging Sumaform's and Salt-Jenkins' Salt Trees
+
+The state trees are merged by creating symbolic links inside sumaform's `salt`
+directory. These symbolic links point to each file in `salt-jenkins/state-tree`
+directory. For directories that exist in both tree, the files inside the
+respective directory in `salt-jenkin/state-tree` are linked instead.
+
+This symbolic link approach has the advantage that it clearly shows where each
+state file is maintained and can be cleaned up easily. Terraform's `file`
+provider, that sumaform uses to upload the `salt/` directory to each VM for
+provisioning can't handle symbolic links well. A
+[workaround](https://github.com/hashicorp/terraform/issues/13202#issuecomment-321578202)
+is possible, but it will touch a core sumaform mechanism.
+
+#### Alternatives
+
+Alternatively, we can copy the files instead. This loses the information where a
+state file is maintained and requires manually updating the copies whenever the
+used submodule revision is updated. Hard links would solve the last issue, I
+haven't yet confirmed that they work with the `file` provider.
+
+## Provisioning Test VMs for GitHub Actions
+
+GitHub Actions use so-called runners, which are VMs running Windows, macOS or
+Linux. It is possible to host your own runners, but we currently neither do nor
+plan to do that. While we can't use sumaform to provision GitHub-hosted runners,
+we can prepare the runner before executing a test suite run. With Salt Toaster,
```

Comment:
One option that may or may not be applicable in this case is to use sumaform in SSH mode - that does not do any VM creation, it only runs the SCP+salt part.

Reply:
It's good to know about the SSH mode, I wasn't aware of that!

For this use case, I don't think it's a good alternative. As far as I know, we would need to add a main.tf just for that to openSUSE/salt. The main.tf would need a module that targets localhost, just to execute a Salt state (which we first need to merge as well). That's a bit complicated for our use case compared to my (ugly) PoC GH Actions jobs definition:
```yaml
jobs:
  build-then-test:
    runs-on: ubuntu-22.04
    steps:
      - name: Check out repository code
        uses: actions/checkout@v3
      - name: Install salt-call
        run: |
          sudo apt-get update
          sudo apt-get install salt-common
      - name: Check out Salt Jenkins
        uses: actions/checkout@v3
        with:
          repository: saltstack/salt-jenkins
          path: salt-jenkins
      - name: Install test dependencies
        working-directory: ./salt-jenkins
        run: sudo salt-call --local --file-root=state-tree state.apply golden-images-provision
      - name: Run tests
        run: nox --session 'pytest-3(coverage=False)' -- tests/pytests/unit/